tests/server_test.php

```diff
@@ -0,0 +1,142 @@
+    public static function function_parameters() {
+        return new external_function_parameters([
+            'vfsStream' => new external_value(
+                    PARAM_TEXT, '', VALUE_REQUIRED),
+        ]);
+    }
+
+    public static function function_returns() {
+        return new external_value(
+                PARAM_BOOL, '', VALUE_REQUIRED);
+    }
+
+    public static function function($params) {}
+}
+
+/**
+ * Loaded REST server test suite.
+ *
+ * @group webservice_loadedrest
+ */
+class webservice_loadedrest_server_testcase extends advanced_testcase {
+    /**
+     * Safely create a mock.
+     *
+     * PHPUnit 6.0.0 removed the legacy getMock() method in favour of
+     * createMock(). Since Moodle 3.1 is stuck on PHPUnit 4.8.x we still need
+     * to support the older version.
+     *
+     * @param string $className
+     *
+     * @return PHPUnit_Framework_MockObject_MockObject|PHPUnit\Framework\MockObject\MockObject
+     */
+    public function createMock($originalClassName) {
+        $oldVersionClass = 'PHPUnit_Runner_Version';
+        $versionClass = class_exists($oldVersionClass)
+                ? $oldVersionClass : 'PHPUnit\Runner\Version';
+
+        $version = $versionClass::series();
+        if (version_compare($version, '5.4', '>=')) {
+            return parent::createMock($originalClassName);
+        } else {
+            // As compatible as possible with createMock() as introduced in
+            // 5.4.0, but without disallowMockingUnknownTypes() as it was
+            // introduced in phpunit/phpunit-mock-object v4.0
+            // (phpunit/phpunit v6.0):
+            // https://github.com/sebastianbergmann/phpunit/blob/5.4.0/src/Framework/TestCase.php#L1510-L1529
+            return $this->getMockBuilder($originalClassName)
+                ->disableOriginalConstructor()
+                ->disableOriginalClone()
+                ->disableArgumentCloning()
+                ->getMock();
+        }
+    }
+
+    public function test_invokes_format_correctly() {
+        $this->resetAfterTest();
+        $this->setAdminUser();
+
+        $vfs = vfsStream::setup('webservice_loadedrest', null, [
+            'input' => 'vfsStream = life saver',
+        ]);
+        /** @var vfsStreamFile $inputfile */
+        $inputfile = $vfs->getChild('input');
+
+        $mockformat = $this->createMock(format::class);
+        $mockformat->expects($this->at(0))
+            ->method('get_name')
+            ->willReturn('mock');
+        $mockformat->expects($this->at(1))
+            ->method('deserialise')
+            ->with($inputfile->getContent())
+            ->willReturn(['vfsStream' => 'life saver']);
+        $mockformat->expects($this->at(2))
+            ->method('send_headers');
+        $mockformat->expects($this->at(3))
+            ->method('send_response');
+
+        $server = new webservice_loadedrest_server_mock($mockformat, $inputfile->url());
+        $server->run();
+    }
+}
```

Comment:
And we need to add some assertions to this test, as there currently are none.

Reply:
The test is intended to cover `server`'s interactions with `format`, which is achieved via the [PHPUnit `InvokedAtIndex` matcher](https://phpunit.de/manual/6.5/en/test-doubles.html#test-doubles.mock-objects.tables.matchers). I can't see any value in adding additional assertions?

Reply:
OK, got it. It's just that when run, PHPUnit doesn't output any sort of status for those - it is like you had no tests whatsoever (no dots, no other notification etc.)